File: 0042-Trapping-Rain-Water.py
class Solution(object):
    def trap(self, inputHeights):
        maxHeight = max(inputHeights)
        numWalls = len(inputHeights)
        totalWaterTrapped = 0

        i = 0
        maxHeightLeft = 0
        while inputHeights[i] < maxHeight:
            maxHeightLeft = max(maxHeightLeft, inputHeights[i])
            totalWaterTrapped += (maxHeightLeft - inputHeights[i])
            i += 1

        j = numWalls - 1
        maxHeightRight = 0
        while inputHeights[j] < maxHeight:
            maxHeightRight = max(maxHeightRight, inputHeights[j])
            totalWaterTrapped += (maxHeightRight - inputHeights[j])
            j -= 1

        if i < j:
            totalWaterTrapped += maxHeight * (j - i) - sum(inputHeights[i:j])

        return totalWaterTrapped

File: 0042-Trapping-Rain-Water.js
var trap = function(height) {
    let maxHeight = height.reduce((a, b) => (Math.max(a, b) || 0));
    let numWalls = height.length;
    let totalWaterTrapped = 0;
    
    let i = 0;
    let maxHeightLeft = 0;
    while(height[i] < maxHeight) {
        maxHeightLeft = Math.max(maxHeightLeft, height[i]);
        totalWaterTrapped += (maxHeightLeft - height[i]);
        i += 1;
    }

    let j = numWalls - 1;
    let maxHeightRight = 0;
    while(height[j] < maxHeight) {
        maxHeightRight = Math.max(maxHeightRight, height[j]);
        totalWaterTrapped += (maxHeightRight - height[j]);
        j -= 1;
    }
    
    while (i < j) {
        totalWaterTrapped += (maxHeight - height[i]);
        i += 1;
    }
    
    return totalWaterTrapped;
};